File: apps/web/lib/vault/onepassword.ts
/**
 * 1Password site grants (opt-in half of V5's fill path).
 *
 * A 1Password login is grantable per host exactly like a local vault item:
 * the control plane writes the stable key `op:<vault>/<item>` into the box's
 * site_grants.json, and `air-vault op-fill` refuses unless the frontmost
 * page's host is listed under it. Nothing here runs unless the owner
 * connected a 1Password account (vault_managers.onepassword.enabled), and
 * the only thing ever read out of the box is item NAMES and ids — never a
 * field value (C18/C19).
 */
import type { SupabaseClient } from "@supabase/supabase-js";
import { command } from "../box/client";
import { asRecord } from "../records";

export const OP_GRANT_PREFIX = "op:";

/** Vault names and item titles as they appear in an `op://` reference: no
 * slashes (they would make the reference ambiguous), no control characters. */
const OP_SEGMENT_RE = /^[^/\\\s][^/\\]{0,62}[^/\\\s]$|^[^/\\\s]$/;

export interface OnePasswordItem {
  /** Stable grant key — `op:<vault>/<item>`, field-independent. */
  id: string;
  vault: string;
  item: string;
  /** The reference prefix the agent completes with a field name. */
  ref_prefix: string;
}

export function opGrantKey(vault: string, item: string): string {
  return `${OP_GRANT_PREFIX}${vault}/${item}`;
}

export function isOpGrantKey(key: string): boolean {
  return parseOpGrantKey(key) !== null;
}

/** Inverse of {@link opGrantKey}; null when the key is not a well-formed one. */
export function parseOpGrantKey(
  key: string
): { vault: string; item: string } | null {
  if (!key.startsWith(OP_GRANT_PREFIX)) return null;
  const rest = key.slice(OP_GRANT_PREFIX.length);
  const slash = rest.indexOf("/");
  if (slash <= 0) return null;
  const vault = rest.slice(0, slash);
  const item = rest.slice(slash + 1);
  if (!OP_SEGMENT_RE.test(vault) || !OP_SEGMENT_RE.test(item)) return null;
  return { vault, item };
}

/** True only when the owner explicitly connected a 1Password account. */
export async function onePasswordConnected(
  supabase: SupabaseClient,
  userId: string
): Promise<boolean> {
  const { data } = await supabase
    .from("vault_managers")
    .select("enabled")
    .eq("user_id", userId)
    .eq("manager", "onepassword")
    .maybeSingle();
  return data?.enabled === true;
}

/**
 * List the owner's 1Password LOGIN items through the box's own CLI. The
 * service-account token never leaves the box (the `air-vault` wrapper reads
 * it out of ~/.hermes/.env into the process env), and `op-list` prints item
 * names and vaults only — no field is resolved.
 */
export async function listOnePasswordLogins(
  boxId: string
): Promise<OnePasswordItem[]> {
  const result = await command(boxId, "air-vault op-list", 60);
  if (result.exitCode !== 0) return [];
  return parseOnePasswordLogins(result.stdout);
}

/** Tolerant parse of `air-vault op-list`; anything odd is dropped. */
export function parseOnePasswordLogins(stdout: string): OnePasswordItem[] {
  let payload: unknown;
  try {
    payload = JSON.parse(stdout);
  } catch {
    return [];
  }
  const entries = asRecord(payload)?.["items"];
  if (!Array.isArray(entries)) return [];
  const items: OnePasswordItem[] = [];
  const seen = new Set<string>();
  for (const entry of entries) {
    const record = asRecord(entry);
    if (!record) continue;
    const item = typeof record["item"] === "string" ? record["item"] : "";
    const vault = typeof record["vault"] === "string" ? record["vault"] : "";
    // Titles or vaults carrying a slash cannot be addressed unambiguously in
    // an op:// reference, so they are not grantable here.
    if (!OP_SEGMENT_RE.test(vault) || !OP_SEGMENT_RE.test(item)) continue;
    const id = opGrantKey(vault, item);
    if (seen.has(id)) continue;
    seen.add(id);
    items.push({ id, vault, item, ref_prefix: `op://${vault}/${item}` });
  }
  return items.sort((a, b) => a.id.localeCompare(b.id));
}

File: infra/template/plugins/air-vault/onepassword.py
"""1Password fill path for ``air-vault op-fill`` — strictly opt-in.

This module is inert unless the owner connected a 1Password account: the
control plane's ``enableManager(..., {manager: "onepassword"})`` is the only
thing that writes ``OP_SERVICE_ACCOUNT_TOKEN`` into the box ``.env``. With no
token, ``require_connected()`` raises before ``op`` is ever spawned, so a box
whose owner never opted in never invokes the CLI and never reads a
credential.

With a token, one field resolves in-process via ``op read``: the token
travels in the child's environment (never argv), the value is captured into a
variable and handed straight to the CDP transport, and nothing but the
reference and the host is ever printed (C18/C19).
"""

from __future__ import annotations

import json
import os
import re
import subprocess
from typing import Dict, Tuple

if __package__ in (None, ""):
    from vault_store import VaultError
else:
    from .vault_store import VaultError

TOKEN_ENV = "OP_SERVICE_ACCOUNT_TOKEN"
OP_BIN = "op"
_TIMEOUT = 30.0

# op://<vault>/<item>/<field> — field may be a section path ("section/label").
_REF_RE = re.compile(r"^op://([^/\s]+)/([^/\s]+)/([^\s]+)$")


def connected() -> bool:
    """True only when the owner connected 1Password (the opt-in gesture)."""
    return bool((os.environ.get(TOKEN_ENV) or "").strip())


def require_connected() -> str:
    if not connected():
        raise VaultError(
            "op_not_connected",
            "1Password not connected — connect it from the Vault tab "
            "(Bring your own manager) before using op-fill",
        )
    return (os.environ.get(TOKEN_ENV) or "").strip()


def parse_ref(ref: str) -> Tuple[str, str, str]:
    """Split an ``op://vault/item/field`` reference; anything else refuses."""
    match = _REF_RE.match((ref or "").strip())
    if not match:
        raise VaultError(
            "bad_op_ref",
            'reference must look like "op://<vault>/<item>/<field>"',
        )
    return match.group(1), match.group(2), match.group(3)


def grant_key(vault: str, item: str) -> str:
    """Stable site-grant key for a 1Password item: ``op:<vault>/<item>``.

    The same key is written by the control plane's grant surface, so a
    1Password login is allowlisted per host exactly like a local item.
    """
    return f"op:{vault}/{item}"


def _child_env(token: str) -> Dict[str, str]:
    env = dict(os.environ)
    env[TOKEN_ENV] = token
    # A service account never has a local desktop/session to fall back on.
    env.pop("OP_SESSION", None)
    return env


def list_logins(token: str) -> list:
    """Item titles, vaults and ids for the connected account's LOGIN items.
    ``op item list`` never returns field values."""
    try:
        result = subprocess.run(
            [OP_BIN, "item", "list", "--categories", "Login", "--format", "json"],
            capture_output=True,
            text=True,
            timeout=_TIMEOUT,
            env=_child_env(token),
            check=False,
        )
    except FileNotFoundError:
        raise VaultError("op_missing", "the 1Password CLI (op) is not installed")
    except subprocess.TimeoutExpired:
        raise VaultError("op_list_failed", "1Password timed out")
    if result.returncode != 0:
        raise VaultError("op_list_failed", "1Password refused the item list")
    try:
        payload = json.loads(result.stdout or "[]")
    except ValueError:
        raise VaultError("op_list_failed", "1Password returned unreadable JSON")
    if not isinstance(payload, list):
        return []
    items = []
    for entry in payload:
        if not isinstance(entry, dict):
            continue
        title = entry.get("title")
        vault = entry.get("vault")
        vault_name = vault.get("name") if isinstance(vault, dict) else None
        if not isinstance(title, str) or not isinstance(vault_name, str):
            continue
        if "/" in title or "/" in vault_name:
            continue  # not addressable as an unambiguous op:// reference
        items.append({
            "vault": vault_name,
            "item": title,
            "grant_key": grant_key(vault_name, title),
            "ref_prefix": f"op://{vault_name}/{title}",
        })
    return sorted(items, key=lambda entry: entry["grant_key"])


def read_value(ref: str, token: str) -> str:
    """Resolve ONE field with ``op read``. The value is returned, never
    printed; stderr is surfaced only as a short failure reason."""
    try:
        result = subprocess.run(
            [OP_BIN, "read", "--no-newline", ref],
            capture_output=True,
            text=True,
            timeout=_TIMEOUT,
            env=_child_env(token),
            check=False,
        )
    except FileNotFoundError:
        raise VaultError("op_missing", "the 1Password CLI (op) is not installed")
    except subprocess.TimeoutExpired:
        raise VaultError("op_read_failed", "1Password timed out")
    if result.returncode != 0:
        reason = (result.stderr or "").strip().splitlines()
        raise VaultError(
            "op_read_failed",
            f"1Password refused: {reason[-1][:200]}" if reason else
            "1Password refused the read",
        )
    value = result.stdout
    if not value:
        raise VaultError("op_field_empty", "1Password returned no value")
    return value
